src/gtc/oir.py

```diff
@@ -120,13 +120,18 @@ class ScalarDecl(Decl):
     pass
 
 
+class LocalScalar(ScalarDecl):
+    pass
+
+
 class Temporary(FieldDecl):
     pass
 
 
 class HorizontalExecution(LocNode):
     body: List[Stmt]
     mask: Optional[Expr]
+    declarations: List[LocalScalar]
 
     @validator("mask")
     def mask_is_boolean_field_expr(cls, v: Optional[Expr]) -> Optional[Expr]:
```

Comment:
Please change in line 153

```
params: List[Decl]
```
to
```
params: List[Union[ScalarDecl,FieldDecl]]
```
and fix the test `params` is only for parameters of the stencil. Temporaries should be added to a `declarations` node if that is needed on the Stencil level.

Reply:
Can we change that in a separate PR? That has nothing to do with the optimizations introduced here, this all works the same way as it has before.

Reply:
Currently difficult due to pydantic Union quirks.

Reply:
I would like to fix it here, because the tests add a temporary to params which is wrong.
Since we anyway, want to move away from pydantic, we can do a temporary hacky check, like

```
class FieldDecl(Decl):
    field_decl=""
```
etc. then the Decls are incompatible with each other.

Reply:
Does not help at all as long as temporary has the same fields as Decl I guess. The only fix that works is deriving the `Temporary` directly from `Decl` and adding an extra field to it. But as this probably only fixes this specific issue, but not the same issue the other way around nor the issue in general, I vote for postponing this until pydantic is replaced and this works without workarounds.

Reply:
yes, that's another comment I made above or below

Reply:
All I care for now is not adding a temporary to `params` and if not in this PR, then in another PR, directly afterwards fix the grammar even if pydantic cannot enforce it.

Reply:
The problem is that pydantic does not only not enforce it, but helps destroying it by its invalid conversions…

Reply:
I added a comment.